### PR TITLE
Fix chat/infotext overlap if many chat lines

### LINKED ISSUE #10364: Chat may overlap with infotext if recent_chat_messages is greater than 8
##### Minetest version
```
5.3.0
```

##### OS / Hardware
Operating system: GNU/Linux

##### Summary
If you increase the setting `recent_chat_messages` to a value greater than 8 or 10, the chat will overlap with the infotext. It's 8 when you have debug mode enabled.

For all practical purposes, this basically means you cannot set this setting higher than 8, otherwise, you get an annoying overlap. This is not much higher than the current value of 6.

##### Steps to reproduce
* Set `recent_chat_messages` to 10
* Start game
* Spam the chat
* Optional: Enable debug mode (Adds an offset to the chat messages)
* Point to a node that has an infotext set

Expected: Infotext and chat do not overlap
Actual: Infotext and chat overlap

### PR DESCRIPTION
**Pull request by Wuzzy2**
Fixes #10364.

Previously, the infotext was placed at a static position exactly 200 pixels from the window top. This is OK under default settings, but breaks when `recent_chat_messages` is 8 or higher.

This PR moves the infotext depending on the value of the `recent_chat_messages` value + 2 additional lines to account for the 2 debug mode lines + 1 additional line as "buffer" for better readability if chat is full.
The higher `recent_chat_messages` is, the lower the infotext will go.

With testing, I have determined the following Y offsets (from the top):

`recent_chat_messages=6` (default): 203 (almost the same as the current offset)
`recent_chat_messages=20` (maximum): 483

So Y position 483 seems reasonable to me in the "worst case". Most modern screens should be good enough. And even if not, the user can always opt to reduce the `recent_chat_messages`.

I also tried with the maximum value with a font size of 24 and the infotext was still visible (screen height = 1080).

## Caveats
Since the infotext offset depends on a setting now, it is theoretically possible that the infotext might land outside the screen for small screens and extreme settings. I'd say if the user wants to screw with GUI options, they sometimes have to expect breakage, but they can always “unbreak” things be reverting to the default.
This caveat is pretty much theoretical anyway, because only small windows would be affected.

## To do

This PR is ready.

## How to test

* Use the settings menu to try out different `recent_chat_messages` values
* Start `devtest` and place a Chest of Everything and point at it
* Check if the infotext is visible and readable
* Fill the chat with text (to hit the `recent_chat_messages` limit
* Check if the infotext is STILL readable

### CODE REVIEW

**Comment by Wuzzy2** [comment]
I noticed that I forgot to take `chat_font_size` into account for the offset. Meaning that the offset will be wrong if the player uses a custom `chat_font_size` (non-zero). I have added code to handle this special case.

This means your testing should also involve changing `chat_font_size`.

File: src/client/gameui.cpp [hunk 1]
```diff
@@ -63,9 +63,18 @@ void GameUI::init()
 
 	// At the middle of the screen
 	// Object infos are shown in this
+	unsigned int chat_height;
+	u16 chat_font_size = g_settings->getU16("chat_font_size");
+	if (chat_font_size != 0) {
+		chat_height = g_fontengine->getFont(chat_font_size, FM_Unspecified)->getDimension(L"").Height;
```
Comment (SmallJoker) [inline src/client/gameui.cpp:69]: Usually the font height is checked using `getDimension("Ay")` to take lower parts into account as well.

File: src/client/gameui.cpp [hunk 2]
```diff
@@ -63,9 +63,18 @@ void GameUI::init()
...
+	} else {
+		chat_height = g_fontengine->getTextHeight();
```
Comment (SmallJoker) [inline src/client/gameui.cpp:71]: Could be generalized if you move this code below the `m_guitext_chat` initialization:

`u16 chat_font_size = m_guitext_chat->getActiveFont()->getDimension(L"Ay").Height;`

Or something like that.

**Comment by Wuzzy2** [comment]
This should be better now.

**Comment by Zughy** [comment]
So, wait, let me understand: instead of creating a bounding box for the chat (see #10372 ), the idea is to move things around when there are too many messages? ROTFL

**Comment by Wuzzy2** [comment]
The infotext box position will be determined on game start, then it will stay there for the rest of the session. Nothing will bounce around at runtime.

**Comment by Wuzzy2** [comment]
Bounding box sadly doesn't solve the de facto limitation of `recent_chat_messages`. If we bound the chat height to avoid collision with infotext, chat will always be cut off after the e.g. 8th line. Thus, `recent_chat_messages` will remain pretty useless.

**Comment by Zughy** [comment]
Edited the previous comment. Still, is it really necessary having a custom `recent_chat_messages` field?

**Comment by Wuzzy2** [comment]
I didn't introduce the `recent_chat_messages` field, but I think it's useful. If you ever were on a busy server with >50 players in which chat scrolls past by at Mach 2, you would agree that more than only 6 chat messages should be visible.

Or are you suggesting to increase the number of visible chat messages, but hardcode the amount (again)?

File: src/client/gameui.cpp [hunk 3]
```diff
@@ -82,6 +71,20 @@ void GameUI::init()
 			chat_font_size, FM_Unspecified));
 	}
 
+	// At the middle of the screen
+	// Object infos are shown in this
+	unsigned int chat_font_height = m_guitext_chat->getActiveFont()->getDimension(L"Ay").Height;
```
Comment (nerzhul) [inline src/client/gameui.cpp:76]: u32

**Review by nerzhul — CHANGES_REQUESTED** [review]

**Comment by Zughy** [comment]
If I had that many players and they actually loved to flood like hell, I'd probably stick with a channel system for the chat. For instance, I remember a famous plugin for Minecraft where, in order to write in global chat, you had to add `!` at the beginning of the message. Also, I've played in the most crowded Minecraft servers and I've never had this problem (also because, filters aside, big servers are usually subdivided in different servers, e.g. via bungeecord where chat is not shared).

In short, the amount of flood shouldn't be something the engine should deal with; therefore the maximum amount of visible messages should be hardcoded

**Comment by Zughy** [comment]
Thinking more about it, this could be useful if someone wants to customise the client. It'll cause a problem when/if the chat becomes editable (like, its position), but till that day, why not? Changed my mind

**Comment by paramat** [comment]
PR seems a good approach.

**Review by sfan5 — APPROVED** [review]

**Comment by SmallJoker** [comment]
The infotext can still be unreadable when the chat messages are broken to multiple lines. However, this is already a good improvement.

**Review by SmallJoker — APPROVED** [review]

**Comment by appgurueu** [comment]
Could you check whether it is possible for the infotext to not be on the screen anymore and log a warning/error then?
